compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -557,12 +558,114 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
+                                    // arguably most common case of `expr.into()`.
+                                    let Some(header) =
+                                        self.infcx.tcx.impl_trait_header(impl_def_id)
+                                    else {
+                                        return;
+                                    };
+                                    let target = header.trait_ref.skip_binder().args.type_at(0);
+                                    let ty = header.trait_ref.skip_binder().args.type_at(1);
+                                    if ty == args.type_at(0) {
+                                        paths.push(format!("<{ty} as Into<{target}>>::into"));
+                                        found = true;
+                                    }
+                                });
+                                if found {
+                                    return;
+                                }
+                            }
+
+                            // We're at the `impl` level, but we want to get the same method we
+                            // called *on this `impl`*, in order to get the right DefId and args.
+                            let Some(assoc) = assocs.filter_by_name_unhygienic(name).next() else {
+                                // The method isn't in this `impl`? Not useful to us then.
+                                return;
+                            };
+                            // Let's ignore the generic params and replace them with `_` in the
+                            // suggested path.
+                            let identity_method = ty::GenericArgs::for_item(
+                                self.infcx.tcx,
+                                assoc.def_id,
+                                |param, _| {
+                                    // We don't want to name the arguments, we just want to give an
+                                    // idea of what the syntax is.
+                                    match param.kind {
+                                        ty::GenericParamDefKind::Lifetime => {
+                                            self.infcx.tcx.lifetimes.re_erased.into()
+                                        }
+                                        ty::GenericParamDefKind::Type { .. } => {
+                                            self.next_ty_var(DUMMY_SP).into()
+                                        }
+                                        ty::GenericParamDefKind::Const { .. } => {
+                                            self.next_const_var(DUMMY_SP).into()
+                                        }
+                                    }
+                                },
+                            );
+                            let mut printer = fmt_printer(self, Namespace::ValueNS);
+                            printer.print_def_path(assoc.def_id, identity_method).unwrap();
```

Comment:
Also, that whole `for_suggestion` hack is not necessary. You should just rebase these substs onto the trait's associated item (which you can access via `assoc_item.trait_item_def_id`), and print that, since you're really just printing a method path.

Reply:
Also, it's probably worthwhile to do this inside of a probe, and instead of doing `can_eq` just make an `ObligationCtxt` and use `eq` so the infer vars are constrained correctly.

Reply:
@compiler-errors I'm still to add the `probe`, but otherwise comments are addressed. I'll do another round of clean up, but I'm intrigued if you think https://github.com/rust-lang/rust/pull/128653/commits/e302e7f90be281eff3c55bb8e0a52e7b121ba883 is worth it.